### PR TITLE
azurerm_storage_account_customer_managed_key - support ManagedHSM Key Vaults

### PR DESCRIPTION
**Pull request by baaym**
# DRAFT
---

This is part 2 of supporting Managed HSM Key Vaults for Storage Account Customer Managed Keys. The previous PR (#19801) only fixed the validation for the planning phase. I noticed that during execution there's also some parsing happening that only supports the regular Key Vault ID. This PR tries to add support for Managed HSM Key Vault IDs.

The reason I kept this in draft is that I'm not sure what the best approach would be. The existing code is tailored towards regular Key Vaults by the use of the `VaultId` struct. Parsing a HSM Key Vault ID results in a different struct (`ManagedHSMId`), even though both are structurally the same. Ideally I would have the `keyVaultParse.VaultID()` and `keyVaultParse.ManagedHSMID()` functions return the same struct, but since it's all generated I don't know how that could be done (if desirable at all).

So the alternative solution I chose now is that I parse the Managed HSM ID and then manually create a `VaultId` from its values, and then the rest of the code is left untouched. Since the API for either Key Vault type is the same this should work (and I'll try and validate it this time by executing a deployment locally using the new binary with this fix). However before it comes to that I would like to know if this approach is supported by the maintainers.

### CODE REVIEW

**Comment by katbyte** [comment]
Hey @baaym,

So it seems like HSM key vaults are a different resource then the regular keyvaults? with a different id (and likely different behaviour and permissions?) - so what we'll have to do is (i think) create the key_vault_hsm_* resources for the HSM vaults (certificate, keys etc) as all the existing keyault ones will not be compatible  and then figure out how to making this resource work with a HSM key. possibly with a different property called key_vault_hsm_id ? 

how does this all fit together in the portal?

**Comment by baaym** [comment]
Hi @katbyte,

It actually appears that the HSM vaults are exactly the same as regular vaults from an API perspective, even though they have slightly different ID's. You can see it [on this microsoft page](https://learn.microsoft.com/en-us/azure/storage/common/customer-managed-keys-overview).

---
image
---

We see it on MSSQL as well, there we can either stick in a regular Key Vault Key ID or a Managed HSM Key Vault Key ID - either works. Also through Terraform btw: https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/mssql/mssql_server_transparent_data_encryption_resource.go#L55

If you look at the actual identifiers you see they're all part of the same resource:
- `/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/REDACTED/providers/Microsoft.KeyVault/managedHSMs/hsm-name`
- `/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/REDACTED/providers/Microsoft.KeyVault/vaults/key-vault-name`

Basically when looking at various documentation both regular and HSM Key Vaults can be used interchangibly when configuring encryption.

That's also why I'm trying to make the a change such that it aligns with the regular Key Vault implementation as much as possible. A Managed HSM Key Vault is just a Key Vault.

**Comment by tombuildsstuff** [comment]
@baaym whilst that's true today, our concern is these end up diverging behaviourally over time, since we've seen this previously with API's which are compatible initially but end up diverging behaviourally (e.g. App Service).

As such whilst there's some benefits to sharing the same nested resources as a the regular Key Vault resource today, we believe it's probably prudent to support this via independent resources, rather than trying to shoe-horn this into the existing resources (and needing to deprecate/ask users to move off those in the future).

This also comes down to complexity within the Key Vault resources, whilst we're planning to support provisioning resources via Private Link, doing so with a Managed HSM doubles the complexity of that resources - to the point where testing this becomes increasingly complicated for each additional combination. Ultimately that's a trade-off, and it's likely that we could essentially duplicate these resources internally reusing much of the same logic - whilst we can refactor that internally later to split them as needed, asking users to migrate across is a larger ask, and is something we generally try to avoid - hence splitting this into multiple sub-resources.

WDYT?

**Comment by baaym** [comment]
Hi @tombuildsstuff,

Thanks for your message. I was writing a very convincing response (😉), but then I remembered that this change is not complete yet. I will complete and test it first, just to have something working before deciding on how to structure it.

**Comment by baaym** [comment]
Hi @tombuildsstuff,

I have gotten it to work (still not intended for production btw), though differently than originally anticipated. You'll see 2 commits, let me explain what I did there.

In the first commit I duplicated the logic to retrieve the Key Vault URI to also work for Managed HSMs. This has been done in quite a rudimentary way, duplicating the logic and replacing it with Managed HSM equivalents. Then in the main resource logic I switch between Key Vault and Managed HSM code paths based on the resource ID (`/vaults/*` vs `/managedHSMs/*`).
That seems to work however I hit a practical issue: I'm not allowed to inspect the Managed HSM. This is a corporate policy. Basically we get assigned a HSM key id that we can access, but we can't read the main HSM to retrieve the URI needed to configure encryption.

This is when I realized why the other encryption resources work transparently between Key Vault and Managed HSMs: they accept this URI as a parameter. Only here in the storage account the Key Vault is accessed by it's Azure Resource ID to retrieve the URI.

So in the second commit I added a new parameter `key_vault_uri`. If it's populated, I skip looking up the URI and directly pass it on to the encryption configuration. This worked and resulted in Storage Account encryption working with the HSM:
Screenshot 2023-02-06 at 12 29 05

So now I'd like to implement this properly, also by directly providing the Key Vault URI instead of looking it up. I have some questions regarding this:

- The Key Vault is now inspected for purge protection/soft delete options, and also to retrieve the URI. Other encryption resources like [mssql_server_transparent_data_encryption_resource](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/mssql/mssql_server_transparent_data_encryption_resource.go) do not do this. Is this logic mandatory? Without this logic the code can be greatly simplified.
- If the URI can become a parameter instead of a lookup, do you think this can be implemented in the existing resource in a backwards-compatible way? Or can this only be implemented in a new resource?

I think there are good arguments for passing on the vault URI directly. If you already have access to the key vault in Terraform, then the URI can be populated by referencing the [vault_uri](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/key_vault#vault_uri) attribute. If you don't have access to the key vault like in our case, then we'll be able to pass it on manually. Furthermore then it works the same as `mssql_server_transparent_data_encryption_resource` for example.

Please let me know what you think!

With kind regards,
Martijn

**Comment by katbyte** [comment]
@baaym, 2 PRs have been opened by microsoft which should start to implement this functionality once merged (#20855 and #20976) - as such i'm going to close this in favour of them. thanks for getting this started!

**Comment by baaym** [comment]
Thanks @katbyte, I'll keep an eye on the new pull requests :+1:
